### PR TITLE
Add typing improvements to improve end-user experience

### PR DESCRIPTION
**Pull request by tylerburdsall**
Addresses:
* https://github.com/WarriorOfWire/goodmetrics_kotlin/issues/3
* https://github.com/WarriorOfWire/goodmetrics_kotlin/issues/4

I suspect there are cleaner ways to go about this for the `sealed interface` implementation, totally open to feedback

### CODE REVIEW

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt [hunk 1]
```diff
@@ -1,5 +1,22 @@
 package goodmetrics
 
+sealed interface SupportedDimensionType
+data class StringDimension(val value: String) : SupportedDimensionType {
+    override fun toString(): String {
+        return value
+    }
```
Comment (tylerburdsall) [inline kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:7]: I overrode these to ensure the intended `value` is written as expected

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt [hunk 2]
```diff
@@ -1,5 +1,22 @@
 package goodmetrics
 
+sealed interface SupportedDimensionType
+data class StringDimension(val value: String) : SupportedDimensionType {
```
Comment (kvcache) [inline kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:4]: How about
```
sealed interface Dimension {
  val name: String

  data class String(val name: String, val value: String) : Dimension
  data class Number(val name: String, val value: Long) : Dimension
  data class B[...]
}
```

which would show up in use a little nicer as `Dimension.String` and whatnot, while things that only care about the interface get to take a `Dimension`.

Could go here or nested inside of `Metrics` to make it `Metrics.Dimension.String` if you prefer

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt (hunk 1 above)
Comment (kvcache) [reply kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:7]: seems reasonable, idk if it's really necessary

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt (hunk 2 above)
Comment (tylerburdsall) [reply kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:4]: Works for me, happy to nest it inside of `Metrics`

File: kotlin/lib/src/main/kotlin/goodmetrics/pipeline/Aggregator.kt [hunk 3]
```diff
@@ -93,16 +96,16 @@ class Aggregator(
     }
 }
 
-typealias DimensionPosition = Set<Map.Entry<String, Any>>
+typealias DimensionPosition = Set<Dimension>
```
[the remark this answers is not in the document]
Comment (tylerburdsall) [reply kotlin/lib/src/main/kotlin/goodmetrics/pipeline/Aggregator.kt:99]: Thank you for the clarification!

File: kotlin/lib/src/main/kotlin/goodmetrics/Client.kt [hunk 4]
```diff
@@ -139,11 +140,11 @@ fun Any.asDimension(): Dimension = when (this) {
...
 }
 
-fun Set<Map.Entry<String, Any>>.initializeDatum(timestampNanos: Long, name: String): Datum = datum {
+fun MetricPosition.initializeDatum(timestampNanos: Long, name: String): Datum = datum {
     unixNanos = timestampNanos
     metric = name
-    for ((metric, position) in this@initializeDatum) {
-        dimensions[metric] = position.asDimension()
+    for (position in this@initializeDatum) {
+        dimensions[metric] = position.value.asDimension()
```
Comment (WarriorOfWire) [inline kotlin/lib/src/main/kotlin/goodmetrics/Client.kt:147]: this needs to be `dimensions[position.name]`. see https://github.com/WarriorOfWire/goodmetrics_kotlin/issues/7 🤦

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt [hunk 5]
```diff
@@ -8,20 +8,27 @@ class Metrics internal constructor(
     internal var timestampMillis: Long,
     internal val startNanoTime: Long,
 ) {
+    sealed interface Dimension {
+        val name: String
+        val value: Any
+        data class StringDimension(override val name: String, override val value: String): Dimension
+        data class NumberDimension(override val name: String, override val value: Long): Dimension
+        data class BooleanDimension(override val name: String, override val value: Boolean): Dimension
```
Comment (WarriorOfWire) [inline kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:16]: nit: these are all already namespaced under `Dimension`. It's pointlessly redundant to repeat Dimension before and after the name. `Dimension.String` is less line noise than `Dimension.StringDimension`

**Review by WarriorOfWire — COMMENTED** [review]
looking really good, just a couple notes and a new issue

File: kotlin/lib/src/main/kotlin/goodmetrics/Client.kt (hunk 4 above)
Comment (tylerburdsall) [reply kotlin/lib/src/main/kotlin/goodmetrics/Client.kt:147]: Thank you for the catch!

File: kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt (hunk 5 above)
Comment (tylerburdsall) [reply kotlin/lib/src/main/kotlin/goodmetrics/Metrics.kt:16]: Right, so if we try to simply name it as `String` the Kotlin compiler will presume all future parameters labeled `String` will be a `Metrics.Dimension.String`, requiring us to explicitly name all `String` parameters in the file as a `kotlin.String`. 

I guess I could always name them to be shorthand, like `Metrics.Dimension.Str` or `Metrics.Dimension.Bool`

**Review by WarriorOfWire — APPROVED** [review]
This is really nice, thanks.

**Pull request merged**